### PR TITLE
Non-uniform-axes - fix attributes in axes type conversion

### PR DESCRIPTION
**Pull request by jlaehne**
### Description of the change
I realized that the conversion function between different axes types (e.g. `convert_to_non_uniform_axis`) do not properly delete the attributes of the previous axes type. Only the dictionary is updated, but the attributes are still present.

1. uniform data axis -> functional or non-uniform data axis: `scale` and `offset` should be removed
2. functional data axis -> non-uniform data axis: `_expression`, `_function`, and `x` should be removed

Also, added missing tests for axes conversion functions.

### Progress of the PR
- [x] Change implemented for case 1,
- [x] Change implemented for case 2,
- [x] add tests,
- [x] ready for review.

### Minimal example of the bug fix or the new feature
```python
import hyperspy.api as hs
import numpy as np
s = hs.signals.Signal1D(np.arange(10))
s.axes_manager.signal_axes[0].convert_to_functional_data_axis(expression='x**2')
s.scale #should fail
s.axes_manager.signal_axes[0].convert_to_non_uniform_axis()
s._expression #should fail
s.x #should fail
```

### CODE REVIEW

**Comment by jlaehne** [comment]
For the attributes which are traits, i.e. `offset` and `scale` for `DataAxis` and `x` for `FunctionalDataAxis`, neither `del self.attribute` nor `self.remove_trait('attribute')` work to remove the attributes from the object. They are just set to their default values. Any suggestions on how to solve that @ericpre or @francisco-dlp ?

**Comment by ericpre** [comment]
See https://github.com/jlaehne/hyperspy/pull/4 to fix the issue.

This behaviour is not documented but it is covered by tests: https://github.com/enthought/traits/issues/1423 and https://github.com/enthought/traits/blob/8ec7fb793278c4706fe34d9096baa56e106f7dc2/traits/tests/test_dynamic_trait_definition.py#L42-L54

**Comment by jlaehne** [comment]
Thanks @ericpre ! I'll check again if anything should be added to the tests and then it should be ready.

**Comment by jlaehne** [comment]
There is some functionality that could be covered by additional tests, but unrelated to this PR, so I will launch a separate PR for that. This one should then be ready.

File: hyperspy/axes.py [hunk 1]
```diff
@@ -1071,17 +1075,23 @@ def convert_to_functional_data_axis(self, expression, units=None, name=None, **k
...
+        self.remove_trait('scale')
+        self.remove_trait('offset')
         self.__class__ = FunctionalDataAxis
         d["_type"] = 'FunctionalDataAxis'
         self.__init__(expression=expression, x=UniformDataAxis(**this_kwargs), **d)
         self.axes_manager = axes_manager
 
     def convert_to_non_uniform_axis(self):
         d = super().get_axis_dictionary()
+        super()._init_trait_listeners()
```
Comment (ericpre) [inline hyperspy/axes.py]: Why is this necessary?

**Review by ericpre — COMMENTED** [review]
I have left one question.

File: hyperspy/axes.py (hunk 1 above)
Comment (jlaehne) [reply hyperspy/axes.py]: Good point, must have remained from one of my tries to get the traits removal working.

Comment (jlaehne) [reply hyperspy/axes.py]: (Now removed)
